File: src/bosh-monitor/lib/bosh/monitor/deployment.rb
module Bosh::Monitor
  class Deployment

    attr_reader :name
    attr_reader :agent_id_to_agent
    attr_reader :instance_id_to_agent
    attr_reader :teams
    attr_writer :locked

    def initialize(deployment_data)
      @logger = Bhm.logger
      @name = deployment_data['name']
      @teams = deployment_data['teams']
      @locked = deployment_data['locked']
      @instance_id_to_instance = {}
      @agent_id_to_agent = {}
      @instance_id_to_agent = {}
    end

    def self.create(deployment_data)
      unless deployment_data.kind_of?(Hash)
        Bhm.logger.error("Invalid format for Deployment data: expected Hash, got #{deployment_data.class}: #{deployment_data}")
        return nil
      end

      unless deployment_data['name']
        Bhm.logger.error("Deployment data has no name: got #{deployment_data}")
        return nil
      end

      Deployment.new(deployment_data)
    end

    def add_instance(instance)
      unless instance
        return false
      end

      instance.deployment = name
      @logger.debug("Discovered new instance #{instance.id}") if @instance_id_to_instance[instance.id].nil?
      @instance_id_to_instance[instance.id] = instance
      true
    end

    def remove_instance(instance_id)
      @instance_id_to_agent.delete(instance_id)
      @instance_id_to_instance.delete(instance_id)
    end

    def instance(instance_id)
      @instance_id_to_instance[instance_id]
    end

    def instances
      @instance_id_to_instance.values
    end

    def instance_ids
      @instance_id_to_instance.keys.to_set
    end

    # Processes VM data from BOSH Director,
    # extracts relevant agent data, wraps it into Agent object
    # and adds it to a list of managed agents.
    def upsert_agent(instance)
      @logger.info("Adding agent #{instance.agent_id} (#{instance.job}/#{instance.id}) to #{name}...")

      agent_id = instance.agent_id

      if agent_id.nil?
        @logger.warn("No agent id for instance #{instance.job}/#{instance.id} in deployment #{name}")
        #count agents for instances with deleted vm, which expect to have vm
        if instance.expects_vm? && !instance.has_vm?
          agent = Agent.new("agent_with_no_vm", deployment: name)
          @instance_id_to_agent[instance.id] = agent
          agent.update_instance(instance)
        end
        return false
      end

      # Idle VMs, we don't care about them, but we still want to track them
      if instance.job.nil?
        @logger.debug("VM with no job found: #{agent_id}")
      end

      agent = @agent_id_to_agent[agent_id]

      if agent.nil?
        @logger.debug("Discovered agent #{agent_id}")
        agent = Agent.new(agent_id, deployment: name)
        @agent_id_to_agent[agent_id] = agent
        @instance_id_to_agent.delete(instance.id) if @instance_id_to_agent[instance.id]
      end

      agent.update_instance(instance)

      true
    end

    def remove_agent(agent_id)
      @agent_id_to_agent.delete(agent_id)
    end

    def agent(agent_id)
      @agent_id_to_agent[agent_id]
    end

    def agents
      @agent_id_to_agent.values
    end

    def agent_ids
      @agent_id_to_agent.keys.to_set
    end

    def update_teams(new_teams)
      @teams = new_teams
    end

    def locked?
      @locked
    end
  end
end

File: src/bosh-monitor/pkg/instance/deployment.go
package instance

import (
	"time"

	"github.com/cloudfoundry/bosh/src/bosh-monitor/pkg/director"
)

type Deployment struct {
	DeploymentName       string
	Teams                []string
	Locked               bool
	instanceIDToInstance map[string]*Instance
	agentIDToAgent       map[string]*Agent
	instanceIDToAgent    map[string]*Agent

	agentTimeout    time.Duration
	rogueAgentAlert time.Duration
}

func NewDeployment(data director.Deployment, agentTimeout, rogueAgentAlert time.Duration) *Deployment {
	return &Deployment{
		DeploymentName:       data.Name,
		Teams:                data.Teams,
		Locked:               data.Locked,
		instanceIDToInstance: make(map[string]*Instance),
		agentIDToAgent:       make(map[string]*Agent),
		instanceIDToAgent:    make(map[string]*Agent),
		agentTimeout:         agentTimeout,
		rogueAgentAlert:      rogueAgentAlert,
	}
}

// CreateDeployment builds a Deployment from a director response, returning nil
// for an entry with no name (which the manager skips).
func CreateDeployment(data director.Deployment, agentTimeout, rogueAgentAlert time.Duration) *Deployment {
	if data.Name == "" {
		return nil
	}
	return NewDeployment(data, agentTimeout, rogueAgentAlert)
}

func (d *Deployment) Name() string {
	return d.DeploymentName
}

func (d *Deployment) AddInstance(inst *Instance) bool {
	if inst == nil {
		return false
	}
	inst.Deployment = d.DeploymentName
	d.instanceIDToInstance[inst.InstanceID] = inst
	return true
}

func (d *Deployment) RemoveInstance(instanceID string) {
	delete(d.instanceIDToAgent, instanceID)
	delete(d.instanceIDToInstance, instanceID)
}

func (d *Deployment) GetInstance(instanceID string) *Instance {
	return d.instanceIDToInstance[instanceID]
}

func (d *Deployment) Instances() []*Instance {
	result := make([]*Instance, 0, len(d.instanceIDToInstance))
	for _, inst := range d.instanceIDToInstance {
		result = append(result, inst)
	}
	return result
}

func (d *Deployment) InstanceIDs() map[string]bool {
	result := make(map[string]bool, len(d.instanceIDToInstance))
	for id := range d.instanceIDToInstance {
		result[id] = true
	}
	return result
}

func (d *Deployment) UpsertAgent(inst *Instance) bool {
	agentID := inst.AgentID
	if agentID == "" {
		if inst.ExpectsVM && !inst.HasVM() {
			agent := NewAgent("agent_with_no_vm",
				WithDeployment(d.DeploymentName),
				WithAgentTimeout(d.agentTimeout),
				WithRogueAgentAlert(d.rogueAgentAlert),
			)
			d.instanceIDToAgent[inst.InstanceID] = agent
			agent.UpdateInstance(inst)
		}
		return false
	}

	agent := d.agentIDToAgent[agentID]
	if agent == nil {
		agent = NewAgent(agentID,
			WithDeployment(d.DeploymentName),
			WithAgentTimeout(d.agentTimeout),
			WithRogueAgentAlert(d.rogueAgentAlert),
		)
		d.agentIDToAgent[agentID] = agent
		delete(d.instanceIDToAgent, inst.InstanceID)
	}
	agent.UpdateInstance(inst)
	return true
}

func (d *Deployment) RemoveAgent(agentID string) {
	delete(d.agentIDToAgent, agentID)
}

func (d *Deployment) GetAgent(agentID string) *Agent {
	return d.agentIDToAgent[agentID]
}

func (d *Deployment) Agents() []*Agent {
	result := make([]*Agent, 0, len(d.agentIDToAgent))
	for _, a := range d.agentIDToAgent {
		result = append(result, a)
	}
	return result
}

func (d *Deployment) AgentIDs() map[string]bool {
	result := make(map[string]bool, len(d.agentIDToAgent))
	for id := range d.agentIDToAgent {
		result[id] = true
	}
	return result
}

func (d *Deployment) AgentIDToAgent() map[string]*Agent {
	return d.agentIDToAgent
}

func (d *Deployment) InstanceIDToAgent() map[string]*Agent {
	return d.instanceIDToAgent
}

func (d *Deployment) UpdateTeams(teams []string) {
	d.Teams = teams
}

func (d *Deployment) IsLocked() bool {
	return d.Locked
}
